File: 2024/day24/go/solution.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Gate struct {
	in1 string
	op  string
	in2 string
	out string
}

func parseInput(filename string) (map[string]int, []Gate, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return nil, nil, err
	}

	parts := strings.Split(strings.TrimSpace(string(content)), "\n\n")
	if len(parts) != 2 {
		return nil, nil, fmt.Errorf("invalid input format")
	}

	// Parse initial wire values
	wires := make(map[string]int)
	for _, line := range strings.Split(parts[0], "\n") {
		parts := strings.Split(line, ": ")
		name := parts[0]
		val, _ := strconv.Atoi(parts[1])
		wires[name] = val
	}

	// Parse gates
	var gates []Gate
	for _, line := range strings.Split(parts[1], "\n") {
		fields := strings.Fields(line)
		// Format: "x00 AND y00 -> z00"
		gates = append(gates, Gate{
			in1: fields[0],
			op:  fields[1],
			in2: fields[2],
			out: fields[4],
		})
	}

	return wires, gates, nil
}

func simulate(wires map[string]int, gates []Gate) map[string]int {
	// Make a copy of wires
	result := make(map[string]int)
	for k, v := range wires {
		result[k] = v
	}

	remaining := make([]Gate, len(gates))
	copy(remaining, gates)

	for len(remaining) > 0 {
		madeProgress := false
		var newRemaining []Gate

		for _, gate := range remaining {
			v1, has1 := result[gate.in1]
			v2, has2 := result[gate.in2]

			if has1 && has2 {
				var output int
				switch gate.op {
				case "AND":
					output = v1 & v2
				case "OR":
					output = v1 | v2
				case "XOR":
					output = v1 ^ v2
				}
				result[gate.out] = output
				madeProgress = true
			} else {
				newRemaining = append(newRemaining, gate)
			}
		}

		remaining = newRemaining
		if !madeProgress && len(remaining) > 0 {
			panic("Circuit stuck - missing inputs")
		}
	}

	return result
}

func getZValue(wires map[string]int) int64 {
	// Get all z wires and sort them
	var zWires []string
	for name := range wires {
		if strings.HasPrefix(name, "z") {
			zWires = append(zWires, name)
		}
	}
	sort.Strings(zWires)

	// Build number from highest to lowest bit
	var result int64
	for i := len(zWires) - 1; i >= 0; i-- {
		result = (result << 1) | int64(wires[zWires[i]])
	}

	return result
}

func part1(wires map[string]int, gates []Gate) int64 {
	finalWires := simulate(wires, gates)
	return getZValue(finalWires)
}

func part2(gates []Gate) string {
	swapped := make(map[string]bool)

	// Build lookup: output -> gate
	gateByOutput := make(map[string]Gate)
	for _, gate := range gates {
		gateByOutput[gate.out] = gate
	}

	// Build lookup: (inputs set, op) -> output
	type InputsOp struct {
		inputs string // sorted comma-separated
		op     string
	}
	gateByInputsOp := make(map[InputsOp]string)
	for _, gate := range gates {
		inputs := []string{gate.in1, gate.in2}
		sort.Strings(inputs)
		key := InputsOp{
			inputs: strings.Join(inputs, ","),
			op:     gate.op,
		}
		gateByInputsOp[key] = gate.out
	}

	// Find the highest bit number
	maxBit := 0
	for _, gate := range gates {
		if strings.HasPrefix(gate.out, "z") {
			bit, _ := strconv.Atoi(gate.out[1:])
			if bit > maxBit {
				maxBit = bit
			}
		}
	}

	maxBitStr := fmt.Sprintf("z%02d", maxBit)

	for _, gate := range gates {
		// Rule: XOR gates that don't take x,y as input should output to z
		if gate.op == "XOR" {
			isXYXor := (strings.HasPrefix(gate.in1, "x") || strings.HasPrefix(gate.in1, "y")) &&
				(strings.HasPrefix(gate.in2, "x") || strings.HasPrefix(gate.in2, "y"))
			if !isXYXor {
				// This is a second-level XOR (sum XOR carry), should output to z
				if !strings.HasPrefix(gate.out, "z") {
					swapped[gate.out] = true
				}
			}
		}

		// Rule: z outputs (except highest bit) should come from XOR
		if strings.HasPrefix(gate.out, "z") && gate.out != maxBitStr {
			if gate.op != "XOR" {
				swapped[gate.out] = true
			}
		}

		// Rule: AND gates (except x00 AND y00) should feed into OR
		if gate.op == "AND" {
			isFirstBit := (gate.in1 == "x00" && gate.in2 == "y00") ||
				(gate.in1 == "y00" && gate.in2 == "x00")
			if !isFirstBit {
				// This AND output should be input to an OR gate
				usedByOr := false
				for _, gate2 := range gates {
					if gate2.op == "OR" && (gate.out == gate2.in1 || gate.out == gate2.in2) {
						usedByOr = true
						break
					}
				}
				if !usedByOr {
					swapped[gate.out] = true
				}
			}
		}

		// Rule: XOR of x,y should feed into another XOR (for z output) or AND (for carry)
		if gate.op == "XOR" {
			isXYXor := (strings.HasPrefix(gate.in1, "x") || strings.HasPrefix(gate.in1, "y")) &&
				(strings.HasPrefix(gate.in2, "x") || strings.HasPrefix(gate.in2, "y"))
			// Skip z00 which is x00 XOR y00 directly
			isZ00 := (gate.in1 == "x00" && gate.in2 == "y00") ||
				(gate.in1 == "y00" && gate.in2 == "x00")
			if isXYXor && !isZ00 {
				// Should be used by XOR and AND
				usedByXor := false
				usedByAnd := false
				for _, gate2 := range gates {
					if gate.out == gate2.in1 || gate.out == gate2.in2 {
						if gate2.op == "XOR" {
							usedByXor = true
						} else if gate2.op == "AND" {
							usedByAnd = true
						}
					}
				}
				if !(usedByXor && usedByAnd) {
					swapped[gate.out] = true
				}
			}
		}
	}

	// Convert to sorted list
	var swappedList []string
	for wire := range swapped {
		swappedList = append(swappedList, wire)
	}
	sort.Strings(swappedList)

	return strings.Join(swappedList, ",")
}

func main() {
	wires, gates, err := parseInput("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(wires, gates))
	fmt.Printf("Part 2: %s\n", part2(gates))
}

File: 2024/day24/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class Solution {

    static class Gate {
        String in1;
        String op;
        String in2;
        String out;

        Gate(String in1, String op, String in2, String out) {
            this.in1 = in1;
            this.op = op;
            this.in2 = in2;
            this.out = out;
        }
    }

    static class InputData {
        Map<String, Integer> wires;
        List<Gate> gates;

        InputData(Map<String, Integer> wires, List<Gate> gates) {
            this.wires = wires;
            this.gates = gates;
        }
    }

    static InputData parseInput(String filename) throws IOException {
        String content = Files.readString(Paths.get(filename));
        String[] parts = content.trim().split("\n\n");

        // Parse initial wire values
        Map<String, Integer> wires = new HashMap<>();
        for (String line : parts[0].split("\n")) {
            String[] tokens = line.split(": ");
            wires.put(tokens[0], Integer.parseInt(tokens[1]));
        }

        // Parse gates
        List<Gate> gates = new ArrayList<>();
        for (String line : parts[1].split("\n")) {
            String[] tokens = line.split(" ");
            gates.add(new Gate(tokens[0], tokens[1], tokens[2], tokens[4]));
        }

        return new InputData(wires, gates);
    }

    static Map<String, Integer> simulate(Map<String, Integer> initialWires, List<Gate> gates) {
        Map<String, Integer> wires = new HashMap<>(initialWires);
        List<Gate> remaining = new ArrayList<>(gates);

        while (!remaining.isEmpty()) {
            boolean madeProgress = false;
            List<Gate> newRemaining = new ArrayList<>();

            for (Gate gate : remaining) {
                if (wires.containsKey(gate.in1) && wires.containsKey(gate.in2)) {
                    int v1 = wires.get(gate.in1);
                    int v2 = wires.get(gate.in2);
                    int result;

                    switch (gate.op) {
                        case "AND":
                            result = v1 & v2;
                            break;
                        case "OR":
                            result = v1 | v2;
                            break;
                        case "XOR":
                            result = v1 ^ v2;
                            break;
                        default:
                            throw new IllegalArgumentException("Unknown operation: " + gate.op);
                    }

                    wires.put(gate.out, result);
                    madeProgress = true;
                } else {
                    newRemaining.add(gate);
                }
            }

            remaining = newRemaining;
            if (!madeProgress && !remaining.isEmpty()) {
                throw new RuntimeException("Circuit stuck - missing inputs");
            }
        }

        return wires;
    }

    static long getZValue(Map<String, Integer> wires) {
        List<String> zWires = new ArrayList<>();
        for (String key : wires.keySet()) {
            if (key.startsWith("z")) {
                zWires.add(key);
            }
        }
        Collections.sort(zWires, Collections.reverseOrder());

        long result = 0;
        for (String z : zWires) {
            result = (result << 1) | wires.get(z);
        }
        return result;
    }

    static long part1(Map<String, Integer> wires, List<Gate> gates) {
        Map<String, Integer> finalWires = simulate(wires, gates);
        return getZValue(finalWires);
    }

    static String part2(List<Gate> gates) {
        Set<String> swapped = new HashSet<>();

        // Build lookup: output -> gate
        Map<String, Gate> gateByOutput = new HashMap<>();
        for (Gate gate : gates) {
            gateByOutput.put(gate.out, gate);
        }

        // Build lookup: (inputs, op) -> output
        Map<String, String> gateByInputsOp = new HashMap<>();
        for (Gate gate : gates) {
            Set<String> inputs = new HashSet<>(Arrays.asList(gate.in1, gate.in2));
            String key = inputs.toString() + ":" + gate.op;
            gateByInputsOp.put(key, gate.out);
        }

        // Find the highest bit number
        int maxBit = 0;
        for (Gate gate : gates) {
            if (gate.out.startsWith("z")) {
                int bit = Integer.parseInt(gate.out.substring(1));
                maxBit = Math.max(maxBit, bit);
            }
        }

        // Check structural rules for a correct adder
        for (Gate gate : gates) {
            // Rule: XOR gates that don't take x,y as input should output to z
            if (gate.op.equals("XOR")) {
                boolean isXyXor = (gate.in1.startsWith("x") || gate.in1.startsWith("y")) &&
                                  (gate.in2.startsWith("x") || gate.in2.startsWith("y"));
                if (!isXyXor) {
                    // Second-level XOR (sum XOR carry) should output to z
                    if (!gate.out.startsWith("z")) {
                        swapped.add(gate.out);
                    }
                }
            }

            // Rule: z outputs (except final carry) should come from XOR
            if (gate.out.startsWith("z") && !gate.out.equals(String.format("z%02d", maxBit))) {
                if (!gate.op.equals("XOR")) {
                    swapped.add(gate.out);
                }
            }

            // Rule: AND gates (except x00 AND y00) should feed into OR
            if (gate.op.equals("AND")) {
                boolean isFirstBit = (gate.in1.equals("x00") && gate.in2.equals("y00")) ||
                                     (gate.in1.equals("y00") && gate.in2.equals("x00"));
                if (!isFirstBit) {
                    // This AND output should be input to an OR gate
                    boolean usedByOr = false;
                    for (Gate g2 : gates) {
                        if (g2.op.equals("OR") && (gate.out.equals(g2.in1) || gate.out.equals(g2.in2))) {
                            usedByOr = true;
                            break;
                        }
                    }
                    if (!usedByOr) {
                        swapped.add(gate.out);
                    }
                }
            }

            // Rule: XOR of x,y should feed into another XOR (for z output) and AND (for carry)
            if (gate.op.equals("XOR")) {
                boolean isXyXor = (gate.in1.startsWith("x") || gate.in1.startsWith("y")) &&
                                  (gate.in2.startsWith("x") || gate.in2.startsWith("y"));
                boolean isZ00 = (gate.in1.equals("x00") && gate.in2.equals("y00")) ||
                                (gate.in1.equals("y00") && gate.in2.equals("x00"));
                if (isXyXor && !isZ00) {
                    boolean usedByXor = false;
                    boolean usedByAnd = false;
                    for (Gate g2 : gates) {
                        if (gate.out.equals(g2.in1) || gate.out.equals(g2.in2)) {
                            if (g2.op.equals("XOR")) {
                                usedByXor = true;
                            } else if (g2.op.equals("AND")) {
                                usedByAnd = true;
                            }
                        }
                    }
                    if (!(usedByXor && usedByAnd)) {
                        swapped.add(gate.out);
                    }
                }
            }
        }

        List<String> result = new ArrayList<>(swapped);
        Collections.sort(result);
        return String.join(",", result);
    }

    public static void main(String[] args) throws IOException {
        InputData data = parseInput("../input.txt");

        System.out.println("Part 1: " + part1(data.wires, data.gates));
        System.out.println("Part 2: " + part2(data.gates));
    }
}
